File: p169e/majority_element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
  let [m, counter] = [null, 0];
  for (let n of nums) {
    if (counter === 0) m = n;
    if (n !== m) counter--;
    else counter++;
  }
  return m;
};
// TEST
[
  [[1], 1],
  [[1, 2, 1], 1],
  [[1, 2, 3, 2, 2], 2],
  [[3, 3, 1, 3, 2, 2, 3], 3]
].forEach(t => {
  const actual = majorityElement(t[0]);
  console.log('Majority element in', t[0], '->', actual);
  console.assert(actual === t[1]);
});

File: p169e/majority_element.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        candidate, counter = None, 0
        for n in nums:
            if counter == 0:
                candidate = n
            counter = counter + 1 if candidate == n else counter - 1
        return candidate


# TESTS
tests = [
    [[1], 1],
    [[1, 2, 1], 1],
    [[1, 2, 3, 2, 2], 2],
    [[3, 3, 1, 3, 2, 2, 3], 3],
]

for t in tests:
    sol = Solution()
    actual = sol.majorityElement(t[0])
    print("Majority element in", t[0], "->", actual)
    assert actual == t[1]
